### PR TITLE
CLI-core tool util: Modify getPluginData to include all data

### PR DESCRIPTION
**Pull request by psealock**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

In an effort to automate version bumping, obtaining the current version and its prerelease data (`-dev`, for instance) will be needed. Previously, Code Analyzer had a function `getPluginData` to get the plugin version stripped of its prerelease data.

This PR changes `getPluginData` to gather all the plugin data, not just the base version.  It now returns all the data in JSON format and moves the function to the core-cli library for easy consumption amongst other tools.

For example, this PHP 

```PHP
<?php
/**
 * Plugin Name: WooCommerce
 * Plugin URI: https://woocommerce.com/
 * Description: An eCommerce toolkit that helps you sell anything. Beautifully.
 * Version: 7.0.0-dev
 * Author: Automattic
 * Author URI: https://woocommerce.com
 * Text Domain: woocommerce
 * Domain Path: /i18n/languages/
 * Requires at least: 5.8
 * Requires PHP: 7.2
 *
 * @package WooCommerce
 */
```
Will be returned as this

```js
{
  pluginName: 'WooCommerce',
  pluginUri: 'https',
  description: 'An eCommerce toolkit that helps you sell anything. Beautifully.',
  version: '6.8.2',
  author: 'Automattic',
  authorUri: 'https',
  textDomain: 'woocommerce',
  domainPath: '/i18n/languages/',
  requiresAtLeast: '5.8',
  requiresPhp: '7.2'
}
```

### How to test the changes in this Pull Request:

1. Add a `console.log( pluginData )` after this line 
https://github.com/woocommerce/woocommerce/blob/9086f59c86aa7c5a308a7bb4afe08b6cfb1ac2ee/tools/code-analyzer/src/commands/analyzer/analyzer-major-minor.ts#L29
2. Run the major-minor command to ensure it still works
```
pnpm -filter=code-analyzer analyzer major-minor -- "release/6.8" "plugins/woocommerce/woocommerce.php"
```
3. See the plugin data logged to the console and the major-minor command works as it did previously

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: tools/cli-core/src/util.ts [hunk 1]
```diff
@@ -170,3 +177,44 @@ export const generateJSONFile = ( filePath: string, data: unknown ) => {
...
+ * @param {string} hashOrBranch   - Hash or branch to checkout.
+ * @return {Object} - plugin data in JSON format.
+ */
+export const getPluginData = async (
+	tmpRepoPath: string,
+	pathToMainFile: string,
+	hashOrBranch: string
+) => {
+	const git = simpleGit( { baseDir: tmpRepoPath } );
+	await git.checkout( [ hashOrBranch ] );
```
Comment (samueljseay) [inline tools/cli-core/src/util.ts]: just a heads up, these lines can be replaced with the helper `await checkoutRef(hashOrBranch)`

File: tools/cli-core/src/util.ts [hunk 2]
```diff
@@ -170,3 +177,44 @@ export const generateJSONFile = ( filePath: string, data: unknown ) => {
...
+
+	const mainFile = join( tmpRepoPath, pathToMainFile );
+
+	const content = await readFile( mainFile, 'utf-8' );
+	const regex = /<\?php\n\/\*\*\n([\S\s]+?) \*\//;
+	const match = content.match( regex );
+
+	if ( ! match || match.length < 2 ) {
+		Logger.error( 'Failed to find plugin data!' );
```
Comment (samueljseay) [inline tools/cli-core/src/util.ts]: As a pattern in helpers and shared utils, don't use the Logger directly. Throw an error instead. Then catch in the workflow (ie when you're consuming it in a Commander action) and use the Logger there. It just gives you much better control over handling error states from workflows.

Consider a situation where you wanted to try getPluginData, but maybe if it failed you didn't want to crash but do something else instead, in that case you would put a try/catch around getPluginData and handle it some other way.

Think of the Logger like the UI layer, it doesn't belong being used in lower levels than the Commander action

File: tools/cli-core/src/util.ts [hunk 3]
```diff
@@ -170,3 +177,44 @@ export const generateJSONFile = ( filePath: string, data: unknown ) => {
...
+		return;
+	}
+
+	return match[ 1 ]
+		.replace( /@package .+\n/, '' )
+		.split( '*' )
+		.reduce( ( result: { [ key: string ]: string }, line: string ) => {
```
Comment (samueljseay) [inline tools/cli-core/src/util.ts]: the type `{ [ key: string ]: string }` can be more simply and easily read typed as `Record<string, string>`. I think TS will still be happy with that. Nitpick but just letting you know :)

File: tools/code-analyzer/src/commands/analyzer/analyzer-major-minor.ts [hunk 4]
```diff
@@ -62,14 +25,17 @@ const program = new Command()
 		const tmpRepoPath = await cloneRepo( source );
 		Logger.endTask();
 
-		const version = await getPluginData(
+		Logger.startTask( `Getting version from ${ pathToMainFile }` );
+		const pluginData = await getPluginData(
```
Comment (samueljseay) [inline tools/code-analyzer/src/commands/analyzer/analyzer-major-minor.ts]: So just to remind you, if you throw instead in `getPluginData` you can do your try catch here and call Logger.error in that case.

**Review by samueljseay — CHANGES_REQUESTED** [review]
This helper looks really sweet and useful! I just have a few small suggestions, probably avoiding Logger.error inside the method is the most important to me, the rest are nitpicks

File: tools/cli-core/src/util.ts (hunk 1 above)
Comment (psealock) [reply tools/cli-core/src/util.ts]: Fixed in https://github.com/woocommerce/woocommerce/pull/34951/commits/4479b1b6ceb9a001b27d37370690202eacde728c

File: tools/cli-core/src/util.ts (hunk 3 above)
Comment (psealock) [reply tools/cli-core/src/util.ts]: Updated in https://github.com/woocommerce/woocommerce/pull/34951/commits/b677f302a202dc1e1bbb93414481f6aec5fb10c7

File: tools/cli-core/src/util.ts (hunk 2 above)
Comment (psealock) [reply tools/cli-core/src/util.ts]: Good call @samueljseay, done in https://github.com/woocommerce/woocommerce/pull/34951/commits/55f8515ddcdd590ef132908e82436b4226c1734f

**Comment by psealock** [comment]
@samueljseay, thanks for all the great feedback. I've addressed the issues you found, could you take another look? 🙇🏽‍♂️

**Comment by psealock** [comment]
Hey @samueljseay, what do you think about this PR? We do need some kind of utility to get the version number as we are doing this several different ways.

**Comment by samueljseay** [comment]
@psealock I'm sorry it looks like this dropped off my radar last year. Is this PR still relevant?

**Pull request closed without merging**